base_exception/models/base_exception.py

```diff
@@ -201,32 +201,58 @@ def detect_exceptions(self):
         """List all exception_ids applied on self
         Exception ids are also written on records
         """
+        all_exception_ids = []  # return of the current def()
```

Comment:
After thinking twice we shlould refactor this code.

He is the proposition

in base_exception module

```
detect_exception(self):
    rules = ....
    for rule in rules:
        records = self._dect_exception(rule)
        reverse_field = self._reverse_exception_field()
        rule.write({reverse_field: [(6, 0, record.ids)]})
```

```
_detect_exception(self, rule):
    if rule.exception_type == 'by_py_code':
        return self._detect_exception_by_py_code(rule)
    elif rule.exception_type == 'by_domain':
        return self._detect_exception_by_py_domain(rule)

def _detect_exception_by_py_code(self, rule):
    ....

def _detect_exception_by_domain(self, rule):
    ....

 ```

Then in sale_exception module

```
class ExceptionRule
    sale_ids = fields.many2many(...)

class SaleOrder

    def _reverse_exception_field()
        return 'sale_ids'

    def detect_exception()
        super(...)
        self.mapped('order_lines).detect_exception()

class SaleOrderLine

    def _reverse_exception_field()
        return 'sale_ids'
    
    def _detect_exception()
        records = super(...)
        return records.mapped('order_id')
```

With that solution we can remove the code "optimize" as it always optimize.
We can remove the sub_exception logic